core/dbt/adapters/base/impl.py

```diff
@@ -1238,13 +1301,52 @@ def calculate_freshness(
 
         snapshotted_at = _utc(table[0][1], source, loaded_at_field)
         age = (snapshotted_at - max_loaded_at).total_seconds()
-        freshness = {
+        freshness: FreshnessResponse = {
             "max_loaded_at": max_loaded_at,
             "snapshotted_at": snapshotted_at,
             "age": age,
         }
         return adapter_response, freshness
 
+    def calculate_freshness_from_metadata(
+        self,
+        source: BaseRelation,
+        manifest: Optional[Manifest] = None,
+    ) -> Tuple[Optional[AdapterResponse], FreshnessResponse]:
+        kwargs: Dict[str, Any] = {
+            "information_schema": source.information_schema_only(),
+            "relations": [source],
+        }
+        result = self.execute_macro(
+            GET_RELATION_LAST_MODIFIED_MACRO_NAME, kwargs=kwargs, manifest=manifest
+        )
+        adapter_response, table = result.response, result.table  # type: ignore[attr-defined]
+
+        try:
+            assert len(table) == 1  # There should be one row since we requested one relation
+            row = table[0]
+            last_modified_val = _get_column_value_uncased("last_modified", row)
+            snapshotted_at_val = _get_column_value_uncased("snapshotted_at", row)
+
+            if last_modified_val is None:
+                # Interpret missing value as "infinitely long ago"
+                max_loaded_at = datetime(1, 1, 1, 0, 0, 0, tzinfo=pytz.UTC)
+            else:
+                max_loaded_at = _utc(last_modified_val, None, "last_modified")
+
+            snapshotted_at = _utc(snapshotted_at_val, None, "snapshotted_at")
+            age = (snapshotted_at - max_loaded_at).total_seconds()
+
+            freshness: FreshnessResponse = {
+                "max_loaded_at": max_loaded_at,
+                "snapshotted_at": snapshotted_at,
+                "age": age,
+            }
+
+            return adapter_response, freshness
+        except Exception:
```

Comment:
This is a long `try` block and a generic `Exception`. Can we limit the `try` block to just the piece that we think will raise the exception, and then also be more specific in the exception that we're catching? For example, I think you're testing for when `len(table) != 1`. Between that and the assert comment, can we move those into an if block and move everything else out of the try block?:

```python
if len(table) != 1:
    raise MacroResultError(...)

row = table[0]
...
```

Reply:
I at least tightened up the scope here.